python/adjoint/utils.py

```diff
@@ -3,14 +3,87 @@
 import meep as mp
 import numpy as onp
 
-from . import ObjectiveQuantity, DesignRegion
+from . import ObjectiveQuantity
 
 # Meep field components used to compute adjoint sensitivities
-_ADJOINT_FIELD_COMPONENTS = [mp.Ex, mp.Ey, mp.Ez]
+_ADJOINT_FIELD_COMPONENTS = [mp.Dx, mp.Dy, mp.Dz]
+_ADJOINT_FIELD_COMPONENTS_CYL = [mp.Dr, mp.Dp, mp.Dz]
 
 # The frequency axis in the array returned by `mp._get_gradient()`
 _GRADIENT_FREQ_AXIS = 1
 
+# The returned axis order from get_dft_array in cylindrical coordinates
+_FREQ_AXIS = 0
+_RHO_AXIS = 2
+_PHI_AXIS = 3
+_Z_AXIS = 1
+
+class DesignRegion(object):
+    def __init__(
+            self,
+            design_parameters,
+            volume=None,
+            size=None,
+            center=mp.Vector3()
+    ):
+        self.volume = volume if volume else mp.Volume(center=center, size=size)
+        self.size = self.volume.size
+        self.center = self.volume.center
+        self.design_parameters = design_parameters
+        self.num_design_params = design_parameters.num_params
+
+    def update_design_parameters(self, design_parameters):
+        self.design_parameters.update_weights(design_parameters)
+    
+    def update_beta(self,beta):
+        self.design_parameters.beta=beta
+
+    def get_gradient(self, sim, fields_a, fields_f, frequencies):
+        num_freqs = onp.array(frequencies).size
+        shapes = []
+        com = _compute_components(sim)
+        scalegrad = 1
+        for ci, c in enumerate(com):
+            '''We need to correct for the rare cases that get_dft_array
+            returns a singleton element for the forward and adjoint fields.
+            This only occurs when we are in 2D and only working with a particular
+            polarization (as the other fields are never stored). Our get_gradient
+            algorithm, however, requires we pass an array of zeros with the
+            proper shape as the design_region.            
+            '''
+            s = sim.get_array_slice_dimensions(c, vol=self.volume)[0]
+            if (onp.squeeze(fields_a[ci][0,...]).size == 1):
+                fields_a[ci] = onp.zeros(onp.insert(s,0,num_freqs))
+                fields_f[ci] = onp.zeros(onp.insert(s,0,num_freqs))
```

Comment:
It would be more readable to do `onp.zeros((num_freqs,) + s)`. Also, would help to use a more descriptive name than `s`... maybe `spatial_shape` or something similar?

Reply:
> maybe spatial_shape or something similar?

Good idea

> onp.zeros((num_freqs,) + s)

unfortunately `s` is a numpy array, and can't be joined like python lists. This seems pretty standard when prepending to numpy arrays, from what I've seen (there are hundreds of possibilities of course...)